flatgfa/src/packedseq.rs

```diff
@@ -0,0 +1,328 @@
+use std::fmt;
+use std::ops::Index;
+
+#[derive(Debug, PartialEq, Eq, Copy, Clone)]
+enum Nucleotide {
+    A,
+    C,
+    T,
+    G,
+}
+
+impl From<u8> for Nucleotide {
+    fn from(value: u8) -> Self {
+        match value {
+            0 => Self::A,
+            1 => Self::C,
+            2 => Self::T,
+            3 => Self::G,
+            4_u8..=u8::MAX => panic!("Not a Nucleotide!"),
+        }
+    }
+}
+
+impl From<Nucleotide> for u8 {
+    fn from(value: Nucleotide) -> Self {
+        match value {
+            Nucleotide::A => 0,
+            Nucleotide::C => 1,
+            Nucleotide::T => 2,
+            Nucleotide::G => 3,
+        }
+    }
+}
+
+impl From<Nucleotide> for char {
+    fn from(value: Nucleotide) -> Self {
+        match value {
+            Nucleotide::A => 'A',
+            Nucleotide::C => 'C',
+            Nucleotide::G => 'G',
+            Nucleotide::T => 'T',
+        }
+    }
+}
+
+/// A compressed vector-like structure for storing nucleotide sequences
+///     - Two base pairs are stored per byte
+///
+/// data: a vector that stores a compressed encoding of this vec's sequence
+/// high_nibble_end: true if the final base pair in the sequence is stored at a
+///                   high nibble
+struct PackedVec {
+    data: Vec<u8>,
+    high_nibble_end: bool,
+}
+
+impl PackedVec {
+    fn new() -> Self {
+        PackedVec {
+            data: Vec::new(),
+            high_nibble_end: true,
+        }
+    }
+    pub fn push(&mut self, input: Nucleotide) {
+        let value = input.into();
+        assert!(value <= 0xF);
+        if self.high_nibble_end {
+            self.data.push(value);
+            self.high_nibble_end = false;
+        } else {
+            let last_index = self.data.len() - 1;
+            self.data[last_index] = (value << 4) | self.data[last_index];
+            self.high_nibble_end = true;
+        }
+    }
+}
+
+impl fmt::Display for PackedVec {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "[")?;
+        let iter = PackedVecIterator::new(&self);
+        let mut i = 0;
+        for item in iter {
+            if i == 0 {
+                i = 1;
+            } else {
+                write!(f, ", ");
+            }
+            println!("u8!!! {}", item);
+            let n: Nucleotide = item.into();
+            let c: char = n.into();
+            println!("CHAR!!! {}", c);
+            write!(f, "{}", c)?;
+        }
+        write!(f, "]")
+    }
+}
+
+struct PackedVecIterator<'a> {
+    data: &'a Vec<u8>,
```

Comment:
I want to suggest a big simplification of the way iterators work. Instead of taking a reference to the underlying `Vec<u8>`, instead take a reference to the entire `PackedVec`. Then, add the following utilities to the `PackedVec` impl:

* `len()`, the number of nucleotides in the vector.
* `get(index: usize)`, which is basically `get_vec_elem`.

Then, this iterator's logic can be substantially simplified. You don't need to keep track of nibble-related issues; you just need to track the current index and stop when you reach `self.vec.len()`. Does this make sense?